_posts/2022-01-13-spice-up-your-javascript-with-curry.md

```diff
@@ -0,0 +1,314 @@
+
+```javascript
+// We need a function that sums up a list of numbers
+const add = curry((a, b) => a + b);
+const reduce = curry((fn, initialValue, list) => list.reduce(fn, initialValue));
+const sum = reduce(add, 0);
+
+// Now combining with the previous example, we have all we need.
+const average = converge(divide, [sum, length]);
+
+const averageSalary = compose(average, pluck("salary"));
+```
+
+## Problem 3: Write functions for gender based average salaries
+
+Still reusing the same functions:
+
+```javascript
+// This is just filtering based by gender and then running the average function
+const filter = curry((fn, list) => list.filter(fn));
+const equals = curry((expected, value) => expected === value);
+const propEq = curry((expectedProp, expectedValue, obj) =>
+  compose(equals(expectedValue), prop(expectedProp))(obj)
+);
+
+const averageGenderSalary = curry((gender, employees) =>
+  compose(averageSalary, filter(propEq("gender", gender)))(employees)
+);
+
+const averageMenSalary = averageGenderSalary("male");
+const averageFemaleSalary = averageGenderSalary("female");
+const averageOtherSalary = averageGenderSalary("other");
+```
+
+## Problem 4 : Write a function that formats an employees name in the form of {firstName} {middleName initials separated by .} {lastName}:
+
+```javascript
+const join = curry((separator, list) => list.join(separator));
+const toUpper = (str) => str.toUpperCase();
+const head = (list) => list[0];
+
+const fullName = compose(
+  join(" "),
+  converge(Array.of, [
+    prop("firstName"),
+    compose(join("."), map(head), prop("middleNames")),
+    prop("lastName"),
+  ])
+);
+```
+
+At this point it is hopefully clear how easy combining functions with composition is.
+
+This type of composition is achieved by having a set of curried functions that can be partially applied as a basic building block.
+
+# Spice... controls the universe? (Pros and cons of curry functions?)
+
+No. This is not the end of all techniques. As with everything there are upsides and downsides, a perfect balance of pros and cons.
+
+There is no silver bullet. This style of _tacit programming_ or _pointfree programming_ requires some time to get used to and in some places, it can get incredibly obscure to read. For people not versed in functional programming these lines of code might seem like some dark invocations. Also, with this style at least from my point of view TypeScript is completely out of the question.
+
+And then there is the worst of them all; debugger support stepping over this kind of code in JavaScript is futile. You are restricted to debugging single functions at a time. However logging as a debugging strategy has existed for a long time and with a simple helper function we can add logging to our composes trivially.
+
+```javascript
+const tap = curry((fn, value) => {
+  fn(value);
+  return value;
+});
+
+const log = tap(console.log);
+
+compose(
+  log, // 5
+  add(1),
+  log, // 4
+  subtract(9)
+)(5);
+```
+
```

Comment:
Maybe "spice... conclusion" to here?

Reply:
Reading the last part again; it definetly falls under a separate header.